### PR TITLE
Overwrite all open file descriptors with /dev/null streams

### PR DESCRIPTION
**Pull request by WyriHaximus**
This also gets rid of STD* streams so we can fully support windows again

Refs https://github.com/reactphp/child-process/pull/65

### CODE REVIEW

**Comment by kelunik** [comment]
Does that work fine on OS X? I think there are portability issues with /proc/self for getting all open FDs IIRC. It also isn't a solution that works on Windows.

**Comment by WyriHaximus** [comment]
@kelunik I'm open for suggestions for windows and OS X 👍

**Comment by Yoshi2889** [comment]
Using `/dev/fd` seems to be the equivalent for `/proc/self/fd` on macOS.

**Comment by WyriHaximus** [comment]
This doesn't work either:
```php
<?php

foreach (new DirectoryIterator('php://fd/*') as $i => $v) {
    var_export([$i, $v]);
}
```
Results in:
> PHP Fatal error:  Uncaught UnexpectedValueException: DirectoryIterator::__construct(php://fd/*): failed to open dir: not implemented in /home/wyrihaximus/Projects/php-file-descriptors/test.php:3

**Comment by kelunik** [comment]
I don't have suggestions, otherwise we would have a solution for Amp already. :-(

**Comment by WyriHaximus** [comment]
@kelunik would a RFC for something like `new DirectoryIterator('php://fd/*')` be doable?

**Comment by WyriHaximus** [comment]
> Not sure, I think php://fd is just a mapping.

Hmm ok, but is or isn't there an internal list of internal file descriptors?

> The correct solution would be defaulting every FD to CLOEXEC.

Would that be hard or easy to do for `7.4` or `8.0`? (That wouldn't help us right now, but having that the default in the future would be awesome.)

**Comment by kelunik** [comment]
There's a list of open resources, but not sure whether there's a list of open file descriptors (including inherited ones).

As PHP is single threaded, closing all resources on proc_open after forking might be good enough.

File: composer.json [hunk 1]
```diff
@@ -10,7 +10,7 @@
         }
     ],
     "require": {
-        "php": "^5.4||^7.0",
+        "php": "^7.2",
```
Comment (WyriHaximus) [inline composer.json]: FYI This will be reverted before merging this to get this fix out to all current users before raising the package to 7.2

**Comment by WyriHaximus** [comment]
> There's a list of open resources, but not sure whether there's a list of open file descriptors (including inherited ones).

Could you look into that? Mainly interested in getting that list on the parent, so we can `null` them.

> As PHP is single threaded, closing all resources on proc_open after forking might be good enough.

Yeah that is working out really well by doing this: https://github.com/WyriHaximus/reactphp-child-process-messenger/pull/22/files#diff-8157a0d28a58887c0293ce14b4b0281fR83

**Comment by kelunik** [comment]
Not sure whether I'll get to that, don't have that much internals knowledge. Maybe @trowski?

File: composer.json (hunk 1 above)
Comment (WyriHaximus) [reply composer.json]: This has been reverted

**Pull request merged**